### PR TITLE
CLI commands

### PR DESCRIPTION
**Pull request by rymurr**
This change is [Reviewable](https://reviewable.io/reviews/projectnessie/nessie/309)

### CODE REVIEW

**Review by jacques-n — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/projectnessie/nessie/309)* status: 0 of 16 files reviewed, 4 unresolved discussions (waiting on @rymurr)

___
*[python/docs/cli.rst, line 33 at r1](https://reviewable.io/reviews/projectnessie/nessie/309#-MJiA8oH1RVb9NZXD0oG:-MJiA8oH1RVb9NZXD0oH:b-1m9e77) ([raw file](https://github.com/projectnessie/nessie/blob/d227d7a38e609db761b91741fc3a96568960ca38/python/docs/cli.rst#L33)):*
> ```reStructuredText
> --------------
> Used to set config parameters found in ``default_config.yaml`` and to set the default context. To set default context use
> ``nessie --add ref main``, all operations that are ref specific will happen on that ref unless otherwise specified.
> ```

Shouldn't this be `nessie config --add ref main`?
___
*[python/docs/cli.rst, line 58 at r1](https://reviewable.io/reviews/projectnessie/nessie/309#-MJia9z52znFk-iVuvMg:-MJia9z52znFk-iVuvMh:b-ivip30) ([raw file](https://github.com/projectnessie/nessie/blob/d227d7a38e609db761b91741fc3a96568960ca38/python/docs/cli.rst#L58)):*
> ```reStructuredText
> .. code-block:: bash
> 
>     Usage: nessie branch [OPTIONS] [BRANCH] [NEW_BRANCH]
> ```

how do I assign a branch to one hash iff it is another hash?
___
*[python/docs/cli.rst, line 65 at r1](https://reviewable.io/reviews/projectnessie/nessie/309#-MJiARrZ5GW1tsPzwYj_:-MJiARrZ5GW1tsPzwYja:brtslej) ([raw file](https://github.com/projectnessie/nessie/blob/d227d7a38e609db761b91741fc3a96568960ca38/python/docs/cli.rst#L65)):*
> ```reStructuredText
>       assign from or rename to
> 
>       examples: nessie branch -l -> list all branches nessie branch -l main ->
> ```

it'd be great if this was formatted better. E.g. bullets with comments or similar.
nessie branch -l # list all branches
nessie branch -l main # list only the main branch (to show hashes)
...
___
*[python/docs/cli.rst, line 84 at r1](https://reviewable.io/reviews/projectnessie/nessie/309#-MJiaCELByfwRwSOVYQq:-MJiaCELByfwRwSOVYQr:b-ed92kp) ([raw file](https://github.com/projectnessie/nessie/blob/d227d7a38e609db761b91741fc3a96568960ca38/python/docs/cli.rst#L84)):*
> ```reStructuredText
> .. code-block:: bash
> 
>     Usage: nessie tag [OPTIONS] [TAG] [NEW_TAG]
> ```

same comments

**Review by rymurr — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/projectnessie/nessie/309)* status: 0 of 16 files reviewed, 4 unresolved discussions (waiting on @jacques-n)

___
*[python/docs/cli.rst, line 33 at r1](https://reviewable.io/reviews/projectnessie/nessie/309#-MJiA8oH1RVb9NZXD0oG:-MJmH6zk-ZRnHmRik1yY:b-896fix) ([raw file](https://github.com/projectnessie/nessie/blob/d227d7a38e609db761b91741fc3a96568960ca38/python/docs/cli.rst#L33)):*
<details><summary><i>Previously, jacques-n (Jacques Nadeau) wrote…</i></summary><blockquote>

Shouldn't this be `nessie config --add ref main`?
</blockquote></details>

Done.
___
*[python/docs/cli.rst, line 58 at r1](https://reviewable.io/reviews/projectnessie/nessie/309#-MJia9z52znFk-iVuvMg:-MJmI7T48qw_BGeKOx3a:b-cwqggy) ([raw file](https://github.com/projectnessie/nessie/blob/d227d7a38e609db761b91741fc3a96568960ca38/python/docs/cli.rst#L58)):*
<details><summary><i>Previously, jacques-n (Jacques Nadeau) wrote…</i></summary><blockquote>

how do I assign a branch to one hash iff it is another hash?
</blockquote></details>

added a condition option
___
*[python/docs/cli.rst, line 65 at r1](https://reviewable.io/reviews/projectnessie/nessie/309#-MJiARrZ5GW1tsPzwYj_:-MJmH9yI4go9gxSb-Ltu:b-896fix) ([raw file](https://github.com/projectnessie/nessie/blob/d227d7a38e609db761b91741fc3a96568960ca38/python/docs/cli.rst#L65)):*
<details><summary><i>Previously, jacques-n (Jacques Nadeau) wrote…</i></summary><blockquote>

it'd be great if this was formatted better. E.g. bullets with comments or similar.
nessie branch -l # list all branches
nessie branch -l main # list only the main branch (to show hashes)
...
</blockquote></details>

Done.
___
*[python/docs/cli.rst, line 84 at r1](https://reviewable.io/reviews/projectnessie/nessie/309#-MJiaCELByfwRwSOVYQq:-MJmHJTX8lD3IAjhT7iU:b-896fix) ([raw file](https://github.com/projectnessie/nessie/blob/d227d7a38e609db761b91741fc3a96568960ca38/python/docs/cli.rst#L84)):*
<details><summary><i>Previously, jacques-n (Jacques Nadeau) wrote…</i></summary><blockquote>

same comments 
</blockquote></details>

Done.

File: python/docs/cli.rst [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+
+    Usage: nessie contents [OPTIONS] [KEY]
+
+      Contents operations.
+
+      KEY name of object to view, delete. If listing the key will limit by
+      namespace what is included.
+
+    Options:
+      -l, --list     list tables
```
Comment (jacques-n) [inline python/docs/cli.rst:227]: we should have an option here to pick the ref we want.

Also, how do I create a key like "foo bar"."my friend"?

File: python/docs/cli.rst [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+
+          nessie branch -f main test -> assign main to head of test
+
+    Options:
+      -l, --list     list branches
+      -d, --delete   delete a branch
+      -f, --force    force branch assignment
+      --json         write output in json format.
+      -v, --verbose  Verbose output.
+      -c, --condition TEXT  Conditional Hash. Only perform the action if branch
```
Comment (jacques-n) [inline python/docs/cli.rst:88]: I would assume this is required except in the cases where force is present. Is that correct?

File: python/docs/cli.rst [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+
+    Commands:
+      add   Set current remote.
+      show  Show current remote.
+
+
+Log Command
+-----------
+
+View the commit log. This operats similarly to ``git log`` and shows the log in the terminals pager. Revision range is
```
Comment (jacques-n) [inline python/docs/cli.rst:156]: operats => operates

File: python/docs/cli.rst [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+specified as <hash>..<hash> or <hash/ref>.
+
+.. code-block:: bash
+
+    Usage: nessie log [OPTIONS] [REVISION_RANGE] [PATHS]...
```
Comment (jacques-n) [inline python/docs/cli.rst:161]: This operation should do the less pagination behavior when in console mode so we can make it paginated later to reduce excessive operations.

File: python/docs/cli.rst [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+branch.
+
+.. code-block:: bash
+
+    Usage: nessie merge [OPTIONS] [MERGE_BRANCH]
+
+      Merge BRANCH into current branch. BRANCH can be a hash or branch.
+
+    Options:
+      -b, --branch TEXT  branch to cherry-pick onto. If not supplied the default
```
Comment (jacques-n) [inline python/docs/cli.rst:190]: should be force or condition.

**Review by jacques-n — COMMENTED** [review]
Looking pretty good. Two main thoughts:

* People need to be able to express an alternative ref versus default somehow for all operations. We should be consistent
* Operations should either require the condition or be a force for conflict safety.

**Review by laurentgo — CHANGES_REQUESTED** [review]
> Reviewed 1 of 11 files at r1, 2 of 15 files at r2, 6 of 21 files at r3.
> *[Reviewable](https://reviewable.io/reviews/projectnessie/nessie/309)* status: 9 of 28 files reviewed, 15 unresolved discussions (waiting on @jacques-n and @rymurr)

___
*[python/docs/cli.rst, line 33 at r1](https://reviewable.io/reviews/projectnessie/nessie/309#-MJiA8oH1RVb9NZXD0oG:-MK6ndUnC8ezdLx2pOVZ:b-ng9w4d) ([raw file](https://github.com/projectnessie/nessie/blob/d227d7a38e609db761b91741fc3a96568960ca38/python/docs/cli.rst#L33)):*
<details><summary><i>Previously, rymurr (Ryan Murray) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

config keys should probably be more explicit (nessie.default-branch), and for some of them (setting default endpoint/default branch), maybe a specific command might be more appropriate. As an example, in git, remote and default branch are changed using git remote 
___
*[python/docs/cli.rst, line 43 at r3](https://reviewable.io/reviews/projectnessie/nessie/309#-MK6oEB7BCYiv482kf85:-MK6oEB7BCYiv482kf86:br1pwd4) ([raw file](https://github.com/projectnessie/nessie/blob/198bb8bf3c72c0e3b23697de6000be9b37505b52/python/docs/cli.rst#L43)):*
> ```reStructuredText
> 
>     Options:
>       --get TEXT    get config parameter
> ```

I quite like git help format: for each command, you get the actual supported format. So instead of 
```
Usage: nessie config [OPTIONS] [KEY]
```

You have:
```
usage: nessie config [<options>] --get <name>
   or: nessie config [<options>] --add <name> <value>
   or: nessie config [<options>] [-l|--list]
 ...

Generic options
  -v | --verbose
...

Specific nessie config actions:
  --get ...
  --add ...
```

Although not all git commands seems to follow the same convention (like git config)

I'm not sure if this file will be left around, or if we plan to have standard man pages or autogenerated help from python/click
___
*[python/docs/cli.rst, line 47 at r3](https://reviewable.io/reviews/projectnessie/nessie/309#-MK6o-Hm6lANOy4oI3oV:-MK6o-Hm6lANOy4oI3oW:b-q7f5hi) ([raw file](https://github.com/projectnessie/nessie/blob/198bb8bf3c72c0e3b23697de6000be9b37505b52/python/docs/cli.rst#L47)):*
> ```reStructuredText
>       -l, --list    list config parameters
>       --unset TEXT  unset config parameter
>       --type TEXT   type to interpret config value to set or get. Allowed options:
> ```

that seems weird to ask use to config how to interpret the config value
___
*[python/docs/cli.rst, line 59 at r3](https://reviewable.io/reviews/projectnessie/nessie/309#-MK6uYJ9EyrmbIpurR9w:-MK6uYJ9EyrmbIpurR9x:b-9rphia) ([raw file](https://github.com/projectnessie/nessie/blob/198bb8bf3c72c0e3b23697de6000be9b37505b52/python/docs/cli.rst#L59)):*
> ```reStructuredText
> .. code-block:: bash
> 
>     Usage: nessie branch [OPTIONS] [BRANCH] [NEW_BRANCH]
> ```

it would be nice to specify a different endpoint
___
*[python/docs/cli.rst, line 74 at r3](https://reviewable.io/reviews/projectnessie/nessie/309#-MK6qD2q5wHn2T059WFZ:-MK6qD2q5wHn2T059WF_:b-4ntmr2) ([raw file](https://github.com/projectnessie/nessie/blob/198bb8bf3c72c0e3b23697de6000be9b37505b52/python/docs/cli.rst#L74)):*
> ```reStructuredText
>           nessie branch -d main -> delete main
> 
>           nessie branch -> list all branches
> ```

It's a bit confusing to support both `nessie branch -l` and `nessie branch` to both support listing, but `nessie branch -l main` and `nessie branch main` are different operations. I guess this is what git does, but something I find confusing
___
*[python/docs/cli.rst, line 80 at r3](https://reviewable.io/reviews/projectnessie/nessie/309#-MK6rUR-AIc2zegfZRwa:-MK6rUR-AIc2zegfZRwb:bby89mk) ([raw file](https://github.com/projectnessie/nessie/blob/198bb8bf3c72c0e3b23697de6000be9b37505b52/python/docs/cli.rst#L80)):*
> ```reStructuredText
>           nessie branch main test -> create branch main at head of test
> 
>           nessie branch -f main test -> assign main to head of test
> ```

how do we assign without forcing? Note that the API for create and assign are separate
___
*[python/docs/cli.rst, line 87 at r3](https://reviewable.io/reviews/projectnessie/nessie/309#-MK6rqXH3RcDgf4ozpWB:-MK6rqXH3RcDgf4ozpWC:b-679gtz) ([raw file](https://github.com/projectnessie/nessie/blob/198bb8bf3c72c0e3b23697de6000be9b37505b52/python/docs/cli.rst#L87)):*
> ```reStructuredText
>       -f, --force    force branch assignment
>       --json         write output in json format.
>       -v, --verbose  Verbose output.
> ```

should be some global option (probably same for json?)
___
*[python/docs/cli.rst, line 235 at r3](https://reviewable.io/reviews/projectnessie/nessie/309#-MK6shwv0X6nXBcYaeNU:-MK6shwv0X6nXBcYaeNV:b-mz2ywi) ([raw file](https://github.com/projectnessie/nessie/blob/198bb8bf3c72c0e3b23697de6000be9b37505b52/python/docs/cli.rst#L235)):*
> ```reStructuredText
>     Options:
>       -l, --list     list tables
>       -d, --delete   delete a table
> ```

missing commit message, unless it opens an editor?

**Review by rymurr — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/projectnessie/nessie/309)* status: 9 of 28 files reviewed, 15 unresolved discussions (waiting on @jacques-n and @laurentgo)

___
*[python/docs/cli.rst, line 33 at r1](https://reviewable.io/reviews/projectnessie/nessie/309#-MJiA8oH1RVb9NZXD0oG:-MK9TsKyBFRAdhrtSmPH:b-xn1jtb) ([raw file](https://github.com/projectnessie/nessie/blob/d227d7a38e609db761b91741fc3a96568960ca38/python/docs/cli.rst#L33)):*
<details><summary><i>Previously, laurentgo (Laurent Goujon) wrote…</i></summary><blockquote>

config keys should probably be more explicit (nessie.default-branch), and for some of them (setting default endpoint/default branch), maybe a specific command might be more appropriate. As an example, in git, remote and default branch are changed using git remote 
</blockquote></details>

I have changed the `ref` to `default_branch`. remote is already changed through `git remote`. I didn't want to use `nessie.default_branch` as I didn't want all config options to be prefixed by a redundant `nessie`.

I didn't want to use remote to change the default branch as to me its not the same as `git remote set-head` and it seems counter intuitive to set it via the remote command. I don't feel too strongly about it so can change if you do.
___
*[python/docs/cli.rst, line 43 at r3](https://reviewable.io/reviews/projectnessie/nessie/309#-MK6oEB7BCYiv482kf85:-MK9UnQ24O55JYNIsVw6:bpbxb4k) ([raw file](https://github.com/projectnessie/nessie/blob/198bb8bf3c72c0e3b23697de6000be9b37505b52/python/docs/cli.rst#L43)):*
<details><summary><i>Previously, laurentgo (Laurent Goujon) wrote…</i></summary><blockquote>

I quite like git help format: for each command, you get the actual supported format. So instead of 
```
Usage: nessie config [OPTIONS] [KEY]
```

You have:
```
usage: nessie config [<options>] --get <name>
   or: nessie config [<options>] --add <name> <value>
   or: nessie config [<options>] [-l|--list]
 ...

Generic options
  -v | --verbose
...

Specific nessie config actions:
  --get ...
  --add ...
```

Although not all git commands seems to follow the same convention (like git config)

I'm not sure if this file will be left around, or if we plan to have standard man pages or autogenerated help from python/click
</blockquote></details>

I plan to set this to generate from click/python. Haven't done it yet (clearly). This file will be published with the rest of the python docs to pynessie.readthedocs.org 

Do you mind if I leave reformatting the help docs for a separate patch? Its definitely nicer the way git does it but it takes a bit of work to generate the help doc by hand. Currently thought the cli will throw an error if used incorrectly eg `nessie config --get x --set x`.
___
*[python/docs/cli.rst, line 47 at r3](https://reviewable.io/reviews/projectnessie/nessie/309#-MK6o-Hm6lANOy4oI3oV:-MK9VSzL4C0yX3V7Yf2L:b-ytsgcx) ([raw file](https://github.com/projectnessie/nessie/blob/198bb8bf3c72c0e3b23697de6000be9b37505b52/python/docs/cli.rst#L47)):*
<details><summary><i>Previously, laurentgo (Laurent Goujon) wrote…</i></summary><blockquote>

that seems weird to ask use to config how to interpret the config value
</blockquote></details>

agreed. Duplicating git in this case. Its a little used but required `git config` option 
___
*[python/docs/cli.rst, line 59 at r3](https://reviewable.io/reviews/projectnessie/nessie/309#-MK6uYJ9EyrmbIpurR9w:-MK9VkxJ3J_5bMSa1xv6:b-5xvy6f) ([raw file](https://github.com/projectnessie/nessie/blob/198bb8bf3c72c0e3b23697de6000be9b37505b52/python/docs/cli.rst#L59)):*
<details><summary><i>Previously, laurentgo (Laurent Goujon) wrote…</i></summary><blockquote>

it would be nice to specify a different endpoint
</blockquote></details>

how come? It seems weird to me to set config but then need to override it everywhere. I would find it weird if git allowed to specify different remote urls depending no how you call it
___
*[python/docs/cli.rst, line 74 at r3](https://reviewable.io/reviews/projectnessie/nessie/309#-MK6qD2q5wHn2T059WFZ:-MK9W84GDFRxcySHPbB_:b-eoumsl) ([raw file](https://github.com/projectnessie/nessie/blob/198bb8bf3c72c0e3b23697de6000be9b37505b52/python/docs/cli.rst#L74)):*
<details><summary><i>Previously, laurentgo (Laurent Goujon) wrote…</i></summary><blockquote>

It's a bit confusing to support both `nessie branch -l` and `nessie branch` to both support listing, but `nessie branch -l main` and `nessie branch main` are different operations. I guess this is what git does, but something I find confusing
</blockquote></details>

Yeah, in this case (and in some above) I have chosen to be as 'git-like' as possible. Git is by no means an ideal or even very consistent UX but my feeling is that if my tool is 90% git but acts differently 10% of the time that 10% will drive me bonkers. I am used to the inconsistencies of the git api so I would rather my new tool have the same inconsistencies. My concern i guess is that if we follow the git api in most places, improving where we want then we will frustrate a percentage of users. If we want to improve over git I would be more inclined to use different commands than to tweak git.
___
*[python/docs/cli.rst, line 80 at r3](https://reviewable.io/reviews/projectnessie/nessie/309#-MK6rUR-AIc2zegfZRwa:-MK9YfzREYhWo8qj90Db:b-4danio) ([raw file](https://github.com/projectnessie/nessie/blob/198bb8bf3c72c0e3b23697de6000be9b37505b52/python/docs/cli.rst#L80)):*
<details><summary><i>Previously, laurentgo (Laurent Goujon) wrote…</i></summary><blockquote>

how do we assign without forcing? Note that the API for create and assign are separate
</blockquote></details>

its weird and again inconsistent. `git branch foo bar` creates foo at bar if foo doesn't exist, otherwise failing. `git branch -f foo bar` creates or reassigns (eg doesn't fail). We are mixing two separate api ops on our side to try and squeeze into git. I am inclined to leave it this way unless you guys disagree  and think we should add an assign command (again to reduce the amount of thinking required by git users).

I have added a `condition` flag so that the assign can be checked but its still not great....I didn't want to add two `force`-like options but thats really what we are saying.
___
*[python/docs/cli.rst, line 87 at r3](https://reviewable.io/reviews/projectnessie/nessie/309#-MK6rqXH3RcDgf4ozpWB:-MK9YDp3-zThB5bfIk4C:b-2har26) ([raw file](https://github.com/projectnessie/nessie/blob/198bb8bf3c72c0e3b23697de6000be9b37505b52/python/docs/cli.rst#L87)):*
<details><summary><i>Previously, laurentgo (Laurent Goujon) wrote…</i></summary><blockquote>

should be some global option (probably same for json?)
</blockquote></details>

true, I have moved json and condition up to global
___
*[python/docs/cli.rst, line 235 at r3](https://reviewable.io/reviews/projectnessie/nessie/309#-MK6shwv0X6nXBcYaeNU:-MK9YPPG2wkiIPyt7kxh:b-n0hl4v) ([raw file](https://github.com/projectnessie/nessie/blob/198bb8bf3c72c0e3b23697de6000be9b37505b52/python/docs/cli.rst#L235)):*
<details><summary><i>Previously, laurentgo (Laurent Goujon) wrote…</i></summary><blockquote>

missing commit message, unless it opens an editor?
</blockquote></details>

have now added `-m` lets you specify a commit message and if you don't it opens an editor

**Comment by rymurr** [comment]
Added global endpoint and `nessie remote set-head` after thinking about your comments @laurentgo

**Review by jacques-n — APPROVED** [review]

**Review by laurentgo — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/projectnessie/nessie/309)* status: 4 of 42 files reviewed, 9 unresolved discussions (waiting on @jacques-n and @laurentgo)

___
*[python/docs/cli.rst, line 43 at r3](https://reviewable.io/reviews/projectnessie/nessie/309#-MK6oEB7BCYiv482kf85:-MKGLOEl0OIMpPxqc4q_:b-jcn358) ([raw file](https://github.com/projectnessie/nessie/blob/198bb8bf3c72c0e3b23697de6000be9b37505b52/python/docs/cli.rst#L43)):*
<details><summary><i>Previously, rymurr (Ryan Murray) wrote…</i></summary><blockquote>

I plan to set this to generate from click/python. Haven't done it yet (clearly). This file will be published with the rest of the python docs to pynessie.readthedocs.org 

Do you mind if I leave reformatting the help docs for a separate patch? Its definitely nicer the way git does it but it takes a bit of work to generate the help doc by hand. Currently thought the cli will throw an error if used incorrectly eg `nessie config --get x --set x`.
</blockquote></details>

Yes, that's fine. I just wanted to make sure it's converging to something consistent :)

**Pull request merged**